File: tests/optest/kernels/78_x_attention/x_attention.cpp
/**
 * Copyright (c) 2026 Huawei Technologies Co., Ltd.
 * This file is a part of the CANN Open Software.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

#include <acl/acl.h>
#include <cmath>
#include <stdexcept>

#include "catlass_kernel_prebuilt.h"
#include "../common/workspace_alloc.h"
#include "x_attention_device.hpp"
#include "x_attention_tiling.hpp"

namespace CatlassKernel {

namespace {

void CheckAcl(aclError status, const char* operation)
{
    if (status != ACL_ERROR_NONE) {
        throw std::runtime_error(std::string(operation) + " failed with ACL error " + std::to_string(status));
    }
}

uint8_t* AllocateWorkspace(size_t size)
{
    if (size == 0) {
        return nullptr;
    }
    if (g_catlassWorkspaceAlloc == nullptr) {
        throw std::runtime_error("CATLASS workspace allocator is not initialized");
    }
    uint8_t* ptr = g_catlassWorkspaceAlloc(size);
    if (ptr == nullptr) {
        throw std::runtime_error("Failed to allocate x_attention workspace");
    }
    return ptr;
}

uint8_t* CopyTilingToDevice(const XAttentionTilingData& tilingData)
{
    if (g_catlassWorkspaceAllocFromHost == nullptr) {
        throw std::runtime_error("CATLASS host-to-device workspace allocator is not initialized");
    }
    uint8_t* ptr = g_catlassWorkspaceAllocFromHost(&tilingData, sizeof(tilingData));
    if (ptr == nullptr) {
        throw std::runtime_error("Failed to copy x_attention tiling data to device");
    }
    return ptr;
}

} // namespace

void XAttention(const uint32_t blockNum, aclrtStream stream, const XAttentionParams& params)
{
    XAttentionTiling::Context context;
    context.batch = params.batch;
    context.beamSize = params.beamSize;
    context.numHeads = params.numHeads;
    context.kvHeads = params.kvHeads;
    context.embeddingSize = params.embeddingSize;
    context.sharedKvSeqLen = params.sharedKvSeqLen;
    context.maxDecodeStep = params.maxDecodeStep;
    context.coreNum = blockNum;
    context.sharedPaged = params.sharedPaged;

    XAttentionTilingData tilingData{};
    uint64_t workspaceSize = 0;
    XAttentionTiling::GetTiling(context, tilingData, workspaceSize);
    if (params.scaleValue > 0.0f) {
        tilingData.scaleValue = params.scaleValue;
    }

    uint8_t* workspace = AllocateWorkspace(workspaceSize);
    uint8_t* tiling = CopyTilingToDevice(tilingData);

    uint64_t hardwareSyncAddr = 0;
    CheckAcl(aclrtGetHardwareSyncAddr(reinterpret_cast<void**>(&hardwareSyncAddr)), "aclrtGetHardwareSyncAddr");

    uint8_t* query = params.inputAddr.at(0);
    uint8_t* sharedKey = params.inputAddr.at(1);
    uint8_t* sharedValue = params.inputAddr.at(2);
    uint8_t* unsharedKey = params.inputAddr.at(3);
    uint8_t* unsharedValue = params.inputAddr.at(4);
    uint8_t* unsharedBlockTable = params.inputAddr.at(5);
    uint8_t* sharedKvLens = params.inputAddr.at(6);
    uint8_t* decodeStep = params.inputAddr.at(7);
    uint8_t* sharedBlockTable = params.inputAddr.at(8);
    uint8_t* output = params.outputAddr.at(0);

    if (params.dataType == ACL_FLOAT16 && params.sharedPaged) {
        ::XAttention<half, true, false><<<blockNum, nullptr, stream>>>(
            hardwareSyncAddr, query, sharedKey, sharedValue, unsharedKey, unsharedValue,
            unsharedBlockTable, sharedKvLens, decodeStep, sharedBlockTable, output, workspace, tiling);
    } else if (params.dataType == ACL_FLOAT16) {
        ::XAttention<half, false, true><<<blockNum, nullptr, stream>>>(
            hardwareSyncAddr, query, sharedKey, sharedValue, unsharedKey, unsharedValue,
            unsharedBlockTable, sharedKvLens, decodeStep, sharedBlockTable, output, workspace, tiling);
    } else if (params.dataType == ACL_BF16 && params.sharedPaged) {
        ::XAttention<bfloat16_t, true, false><<<blockNum, nullptr, stream>>>(
            hardwareSyncAddr, query, sharedKey, sharedValue, unsharedKey, unsharedValue,
            unsharedBlockTable, sharedKvLens, decodeStep, sharedBlockTable, output, workspace, tiling);
    } else if (params.dataType == ACL_BF16) {
        ::XAttention<bfloat16_t, false, true><<<blockNum, nullptr, stream>>>(
            hardwareSyncAddr, query, sharedKey, sharedValue, unsharedKey, unsharedValue,
            unsharedBlockTable, sharedKvLens, decodeStep, sharedBlockTable, output, workspace, tiling);
    } else {
        throw std::runtime_error("x_attention supports float16 and bfloat16 only");
    }
}

} // namespace CatlassKernel

File: tests/optest/torch_catlass/ops/x_attention.py
# This program is free software, you can redistribute it and/or modify.
# Copyright (c) 2026 Huawei Technologies Co., Ltd.
# This file is a part of the CANN Open Software.
# Licensed under CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance
# with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS
# OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.

from typing import Optional

import torch
from torch import Tensor


def _optional_block_table(value: Optional[Tensor], query: Tensor) -> Tensor:
    if value is None:
        return torch.empty(0, dtype=torch.int32, device=query.device)
    return value


def x_attention(
    query: Tensor,
    shared_key_block: Tensor,
    shared_value_block: Tensor,
    unshared_key_block: Tensor,
    unshared_value_block: Tensor,
    unshared_block_table: Optional[Tensor],
    shared_kv_lens: Tensor,
    decode_step: Tensor,
    shared_block_table: Optional[Tensor],
    scale_value: float = 0.0,
) -> Tensor:
    """Run CATLASS XAttention on NPU tensors.

    Source: example 78_x_attention. XAttention combines a batch-shared KV
    context with a beam-private decode cache in one softmax.

    Exactly one cache is paged:

    - ``shared_block_table`` is set and ``unshared_block_table`` is ``None``:
      shared paged cache plus unshared continuous cache.
    - ``shared_block_table`` is ``None`` and ``unshared_block_table`` is set:
      shared continuous cache plus unshared paged cache.

    Contract (see examples/78_x_attention/README.md): positive batch, beam_size,
    num_heads and kv_heads; num_heads % kv_heads == 0; GQA group size <= 128;
    head dimension == 128; 1 <= decode_step <= max_decode_step <= 256.
    All Q/K/V use the same float16 or bfloat16 dtype. All input tensors must
    be contiguous and on the same NPU. Attention masks are not supported.
    The 128-token block size applies only to the shared paged cache; the
    unshared paged cache is indexed by request and uses max_decode_step capacity.
    Length and block-index tensor values must be valid for the cache capacity;
    the wrapper does not check these device-resident values on the host.

    Args:
        query: Query in TND layout ``(batch * beam_size, num_heads, 128)``.
        shared_key_block: Shared key cache. Paged shape is
            ``(num_blocks, 128, kv_heads, 128)``; continuous shape is
            ``(batch * shared_kv_seq_len, kv_heads, 128)``.
        shared_value_block: Shared value cache with the same shape as the shared key cache.
        unshared_key_block: Beam-private key cache. Continuous shape is
            ``(batch * beam_size, kv_heads, max_decode_step, 128)``; paged shape is
            ``(request_count, beam_size, kv_heads, max_decode_step, 128)``.
        unshared_value_block: Unshared value cache with the same shape as the unshared key cache.
        unshared_block_table: Request index for each batch in unshared-paged mode, otherwise ``None``.
        shared_kv_lens: Positive actual shared KV length for each batch, int32
            shape ``(batch,)``. Paged lengths must not exceed table capacity.
            Continuous batches are packed by these lengths; example/ATK use
            equal lengths matching shared_kv_seq_len, without inter-batch padding.
        decode_step: Number of valid unshared cache tokens, int32 shape ``(1,)``;
            one value shared by all batches/beams, in ``[1, max_decode_step]``.
        shared_block_table: Shared paged-cache block table, int32 shape
            ``(batch, max_blocks_per_batch)``, otherwise ``None``.
        scale_value: Positive attention scale, or ``0.0`` for ``1 / sqrt(128)``.

    Returns:
        Tensor with the same shape and dtype as ``query``.
    """
    unshared_table = _optional_block_table(unshared_block_table, query)
    shared_table = _optional_block_table(shared_block_table, query)
    return torch.ops.catlass.x_attention(
        query,
        shared_key_block,
        shared_value_block,
        unshared_key_block,
        unshared_value_block,
        unshared_table,
        shared_kv_lens,
        decode_step,
        shared_table,
        scale_value,
    )
